src/library/scala/concurrent/Future.scala

```diff
@@ -608,9 +608,13 @@ object Future {
    * @return          the `Future` of the `TraversableOnce` of results
    */
   def sequence[A, M[X] <: TraversableOnce[X]](in: M[Future[A]])(implicit cbf: CanBuildFrom[M[Future[A]], A, M[A]], executor: ExecutionContext): Future[M[A]] = {
-    in.foldLeft(successful(cbf(in))) {
-      (fr, fa) => for (r <- fr; a <- fa) yield (r += a)
-    } map (_.result())
+    if (in.isEmpty) {
+      successful(cbf().result())
+    } else {
+      in.foldLeft(successful(cbf(in))) {
+        (fr, fa) => for (r <- fr; a <- fa) yield (r += a)
+      } map (_.result())
```

Comment:
An alternative solution would be to explicitly provide the `InternalCallbackExecutor` to the `map` operation here, which in the case of an empty `in` would be executed on the calling thread and on the thread of the last completing Future (the provided `executor`)